oeps/oep-0028-content-theming-in-react.rst

```diff
@@ -0,0 +1,246 @@
+
+.. code-block:: js
+
+    const Header = (props) => {
+        return (
+            <h1>{props.title}</h1>
+            {props.children}
+        )
+    }
+
+    // use
+    <Header title={"Open Edx"} >
+        <h4>Open Source MOOC platform</h4>
+    </Header>
+
+Another example where specialized children are passed
+
+.. code-block:: js
+
+    const Pane = (props) => {
+        return (
+            <div className="pane">
+                <div className="pane-left">
+                    {props.left}
+                </div>
+                <div className="pane-right">
+                    {props.right}
+                </div>
+            </div>
+
+        )
+    }
+
+    // use
+    <Pane left={<Image />} right={<Explanation />} />
+
+
+* We will use function and placeholder to add additional content in customizable components. We can take an example of the above DefaultTheme and see _MainNav where it has support to add additional nav links by overriding `extraNavLinks` function.
+
+.. code-block:: js
+
+    // Customizable Main Navigation Area
+    class MyThemedNav extends _MainNav {
+        get extraNavLinks() {
+            return [
+                <a href="/about">About Us</a>,
+            ];
+        }
+    }
+
+    // Custom theme:
+    export const MainNav = MyThemedNav;
+    export const MainNavWrapper = _MainNavWrapper;
+
+
+* We will provide support for a global redux store which will be Open edX global redux store. This will act as a central place to store data and will have access to all data available to the system.
+
+* We will consider the layout of the data in the Open edX global redux store as a stable API. We will provide support to pre-fill the store with some common data like current user, current course, list of courses enrolled, etc. We will provide the choice for themes to fetch data that's not part of the Open edX global redux store from REST API's using built-in redux actions and store it in their own separate redux store. We will announce breaking changes if the layout of the data changes in global store.
+
+* We will use containers [2] to access data from the redux store and provide it to components via props. A container is a react component that has a direct connection to the state managed by redux and access data from the state via mapStateToProps. We will use Container as a mechanism to separate data access functionality from the Component. This way we can keep both non redux connected version as well as redux connected version of the same component.
+
+* We will have support to convert any component into a container if it needs to access any data from the redux store, which it currently does not have access to. We can see this by an example where NavbarHeader component initially displays site title. This component now needs to display authenticated username, which is there in the redux store.
```

Comment:
I feel like the example below should be showing a NavbarHeader component that doesn't use the username, and an inherited version that does. The example here doesn't work without the container, so it isn't an example of "converting any component into a container".

Reply:
I have updated this to include `NavbarHeaderWithUserName` which extends from `NavbarHeader` but completely overrides its render function. The extension will help in propagating other functionalities from `NavbarHeader` to `NavbarHeaderWithUserName`.
`NavbarHeader` contains support to display username which is overridden by a container. This also brings us to a point where sometimes both the component(to display additional properties) and container needs to be overridden in case we need to display some additional data.

Reply:
I also don't like the idea of "converting" we are not really converting anything, we are just adding another component that passes some props by default by fetching that data from the store. So it's standard component composition, you're wrapping an existing component with another.

Reply:
@xitij2000 : I have updated the "converting" part to "composition" now. Thanks:)